lib/Sema/TypeCheckProtocol.cpp

```diff
@@ -818,6 +828,10 @@ matchWitness(TypeChecker &tc,
     // Check that the mutating bit is ok.
     if (checkMutating(funcReq, funcWitness, funcWitness))
       return RequirementMatch(witness, MatchKind::MutatingConflict);
+    if (funcWitness->isNonMutating() && funcReq->isConsuming())
+      return RequirementMatch(witness, MatchKind::NonMutatingConflict);
+    if (funcWitness->isConsuming() && !funcReq->isConsuming())
```

Comment:
Sure we can thunk to the owned witness, but that would create a situation where protocol requirements (mistakenly, IMO) marked `consuming` for copyable witness types would automatically regress performance.